src/main/java/de/filefighter/rest/domain/filesystem/business/FileSystemBusinessService.java

```diff
@@ -96,11 +94,45 @@ public void deleteFileSystemItemById(long fsItemId, User authenticatedUser) {
             throw new FileSystemItemCouldNotBeDeletedException(fsItemId);
 
         if (fileSystemEntity.isFile()) {
-            // CASE 0 - File
             fileSystemRepository.delete(fileSystemEntity);
         } else {
-            // CASE 1 - Folder
-
+            HashSet<FileSystemEntity> foundFolders = new HashSet<>();
+            foundFolders.add(fileSystemEntity);
+            Iterator<FileSystemEntity> iterator = foundFolders.iterator();
+
+            // Lists for change or deletion
+            ArrayList<FileSystemEntity> entitiesToBeDeleted = new ArrayList<>();
+            ArrayList<FileSystemEntity> entitiesToBeChanged = new ArrayList<>();
+
+            while (iterator.hasNext()) {
+                FileSystemEntity nextFolder = iterator.next();
+                // no permissions set for comparison.
+                ArrayList<FileSystemEntity> foundEntities = (ArrayList<FileSystemEntity>) getFolderContentsOfEntityAndPermissions(nextFolder, authenticatedUser, false, false);
+                int countOfChildEntities = foundEntities.size();
+                int countOfDeletedEntities = 0;
+
+                for (FileSystemEntity fileSystemEntityToBeDeleted : foundEntities) {
+                    // check here for permissions.
+                    if (userIsAllowedToEditFileSystemEntity(fileSystemEntityToBeDeleted, authenticatedUser) && userIsAllowedToSeeFileSystemEntity(fileSystemEntityToBeDeleted, authenticatedUser)) {
```

Comment:
not visible files/folders can be ignored (just act as if they were deleted)

Reply:
na we cant. I already implemented a check, wheter or not the folder is actually empty after deleting all deletable files.

1. File is visible and editable -> delete it. (folder can be removed.)
2. File is visible and NOT editable -> leave it as it is. (Parent folder cannot be deleted.)
3. File is NOT visible and NOT editable. -> cannot be deleted but folder might appear to be invisible. (Parent folder cannot be removed.)

Reply:
Ah forgot. With 3. you have to change the permissions of the main folder

Reply:
yes.